dist/index.js

```diff
@@ -5,27 +5,30 @@ export default function useAppState(settings) {
   const { onChange, onForeground, onBackground } = settings || {};
   const [appState, setAppState] = useState(AppState.currentState);
 
-  function handleAppStateChange(nextAppState) {
-    if (nextAppState === 'active') {
-      isValidFunction(onForeground) && onForeground();
-    } else if(nextAppState.match(/inactive|background/)) {
-      isValidFunction(onBackground) && onBackground();
-    }
-    setAppState(nextAppState);
-    isValidFunction(onChange) && onChange(nextAppState);
-  }
-
   // didMount effect
   useEffect(() => {
+    function handleAppStateChange(nextAppState) {
+      if (nextAppState === 'active') {
+        isValidFunction(onForeground) && onForeground();
+      } else if (
+        appState === 'active' &&
+        nextAppState.match(/inactive|background/)
+      ) {
+        // to ensure that onBackground() is not called twice we
+        // check if the previous app state was 'active'
+        // as iOS has two not-active-states (inactive and background)
+        isValidFunction(onBackground) && onBackground();
+      }
+      setAppState(nextAppState);
+      isValidFunction(onChange) && onChange(nextAppState);
+    }
     AppState.addEventListener('change', handleAppStateChange);
     // didUnmount effect
     return () => AppState.removeEventListener('change', handleAppStateChange);
-  },[]);
-
+  }, [onChange, onForeground, onBackground, appState]);
```

Comment:
`useEffect` is used here to handle `componentDidMount` and `componentWillUnmount` only and we did that by having empty array in `useEffect` 

In new changes we are suggesting adding all props and state, i wonder why we need to do that  ? 
1. I don't think there is a use case where we need to update the callback functions right ? 
2. By adding props and state in the array we make `useEffect` work now as `componentDidUpdate` right ? this means `useEffect` code will be executed with every update, and this means we will keep adding more listeners right ? not sure if this is right


Please correct me if i am wrong, just trying to understand the changes

Reply:
First of all: I am not saying I understand to 1000% why things are supposed to be a certain way but I am trying to follow the recommendations and the linting rules of the people that built React as closely as possible as they are very likely mostly there for a reason.

First I moved the function inside because that's what linting and documentation recommend to do (as seen here: https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies):
```
It’s difficult to remember which props or state are used by functions outside of the effect. 
This is why usually you’ll want to declare functions needed by an effect inside of it. 
Then it’s easy to see what values from the component scope that effect depends on.
```
Once this function is moved then documentation and linting suggest to (https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects):
```
make sure the array includes all values from the component scope (such as props and state)
that change over time and that are used by the effect. 
Otherwise, your code will reference stale values from previous renders
```

So regarding your points:
- `I don't think there is a use case where we need to update the callback functions right ?`
  There might be and if there is, this code might fail from my understanding. 
- `By adding props and state in the array we make useEffect work now as componentDidUpdate right ? this means useEffect code will be executed with every update, and this means we will keep adding more listeners right ? not sure if this is right` 
  Yes and No. The effect will run more often (whenever `appState` changes pretty much) but as documentation explains: `Effects run for every render and not just once. This is why React also cleans up effects from the previous render before running the effects next time`. So essentially every time `appState` changes we are removing and then re-adding the event listener. And that might seem like a lot of "work" but the experts say it's actually negligible and results in fewer bugs and unexpected behavior (keeping in mind that `useEffect` happens after DOM renders and is therefore non-blocking for the UI).

If you really want to deep-dive, I suggest: https://overreacted.io/a-complete-guide-to-useeffect/ 😅

Reply:
I missed the fact that once you added state and props to the array of `useEffect` unsubscribe code will be also called on each update and as a result we should be good assuming all of these listeners subscriptions and removal is negligible as in React documentation.